presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -401,11 +402,31 @@ protected ExtractionResult visitComparisonExpression(ComparisonExpression node,
             Type fieldType = checkedTypeLookup(symbol);
             NullableValue value = normalized.getValue();
 
-            Optional<NullableValue> coercedValue = coerce(value, fieldType);
-            if (coercedValue.isPresent()) {
-                return createComparisonExtractionResult(normalized.getComparisonType(), symbol, fieldType, coercedValue.get().getValue(), complement);
+            if (!normalized.wasNonCoercionCastRemoved()) {
```

Comment:
My sense is that the comparison expressions should have the same type, if the NULLs are not being CAST, that is a bug in the planner.